File: 673. Number of Longest Increasing Subsequence/673.py
class Solution:
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        """
        len[k] is the length of longest subsequence ending with nums[k];
        cnt[k] is total number of longest subsequence ending with nums[k];

        len[k] = max(len[k], len[i] + 1) for all i <= k and nums[i] < nums[k]
        cnt[k] = sum(cnt[i]) for all i <= k and nums[i] < nums[k] and len[i] = len[k] - 1
        """



        dpLen = [1 for i in range(len(nums))]
        dpCnt = [1 for i in range(len(nums))]
        maxLen, rt = 1, 0
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    if dpLen[i] < dpLen[j] + 1:
                        dpLen[i] = dpLen[j] + 1
                        dpCnt[i] = dpCnt[j]
                    elif dpLen[i] == dpLen[j] + 1:
                        dpCnt[i] += dpCnt[j]
            maxLen = max(maxLen, dpLen[i])

        for i in range(len(nums)):
            if dpLen[i] == maxLen:
                rt += dpCnt[i]
        return rt

if __name__ == "__main__":
    s = Solution()
    nums = [1,3,5,4,7]
    print(s.findNumberOfLIS(nums))
    nums = [2,2,2,2,2]
    print(s.findNumberOfLIS(nums))

File: 673. Number of Longest Increasing Subsequence/673.cpp
#include <vector>
#include <numeric>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution {
public:
    int findNumberOfLIS(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        vector<int> dpLen(nums.size(), 1);              // length of LIS ending with nums[i]
        vector<int> dpCnt(nums.size(), 1);              // count of LIS ending with nums[i]

        for (int i = 1; i < nums.size(); i++) {         // sequence ending with nums[i]
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    if (dpLen[j] + 1 > dpLen[i]) {
                        dpLen[i] = dpLen[j] + 1;
                        dpCnt[i] = dpCnt[j];
                    } else if (dpLen[j] + 1 == dpLen[i]) {
                        dpCnt[i] += dpCnt[j];
                    }
                }
            }
        }

        int result = 0;
        int maxLen = *max_element(dpLen.begin(), dpLen.end());
        for (int i = 0; i < nums.size(); i++) {
            if (dpLen[i] == maxLen) {
                result += dpCnt[i];
            }
        }
        return result;
    }
};

int main() {
    Solution s;
    vector<int> nums = {1,3,5,4,7};
    cout << s.findNumberOfLIS(nums) << endl;
    nums = {2,2,2,2,2};
    cout << s.findNumberOfLIS(nums) << endl;
}
